File: typescript/20_modules.ts
// ============================================================
// Topic:  Modules & Imports  (FOCUS: ES module syntax, import/export, stdlib)
// Summary: export functions, import type, import * as alias, re-export pattern in comments.
// Run:     npx ts-node typescript/20_modules.ts
//
// NOTE: Single-file demo — exports live in this file; multi-file projects split helpers.
//       Circular imports (import top / export bottom of same file) are avoided here.
//
// Re-export patterns (multi-file):
//   export { greet, PI, sqrt } from './helpers';
//   export * from './utils';
//   export type { User } from './types';
//
// KEY DIFFERENCES:
//   Python:  import module / from module import name
//   TS/JS:   import { x } from 'module' / import * as alias / export
//   Rust:    mod + use crate::path
//   Java:    import / import static; java.lang auto-imported
//   C++:     #include <header>
// ============================================================

import { readFileSync } from "node:fs";
import * as alias from "node:fs";
import type { Stats } from "node:fs";
import * as path from "node:path";

// ----- Local exports (used by main below) -----
export const PI = 3.14;

export function greet(name: string): string {
  return "Hi";
}

export function sqrt(n: number): number {
  return Math.sqrt(n);
}

function main(): void {
  // ----- 1. Named export — PI -----
  console.log(`pi: ${PI.toFixed(2)}`); // Output: pi: 3.14

  // ----- 2. Local exported helper — greet -----
  console.log(`greet: ${greet("World")}`); // Output: greet: Hi

  // ----- 3. Exported util — sqrt -----
  console.log(`sqrt: ${sqrt(25.0).toFixed(1)}`); // Output: sqrt: 5.0

  // ----- 4. Namespace import * as alias — Map scores -----
  const scores = new Map<string, number>();
  scores.set("alice", 25);
  console.log(`alias: ${scores.get("alice")}`); // Output: alias: 25

  // ----- 5. Stdlib imports — readFileSync + alias namespace + import type -----
  const tsconfigPath = path.join(__dirname, "tsconfig.json");
  readFileSync(tsconfigPath, "utf8");
  const stats: Stats = alias.statSync(__filename);
  void stats;

  console.log("main module"); // Output: main module
}

main();

// Cross-language note:
//   import { readFileSync } from 'node:fs'  == Python from os import read
//   import * as alias from 'node:fs'        == Python import os as alias
//   import type { Stats }                   == type-only import (erased at compile)
//   export const / export function          == Python module-level public names

File: java/20_modules.java
// ============================================================
// Topic:  Modules & Imports  (FOCUS: java.util, static import, java.lang implicit)
// Summary: import java.util.*, static Math.PI, package note, java.lang implicit.
// Run:     java java/20_modules.java   (Java 11+ single-file launch)
//
// NOTE: Single-file launch uses the default (unnamed) package — no package declaration.
//       Multi-file projects use: package com.example.app;
//
// KEY DIFFERENCES:
//   Python:  import module / from module import name
//   TS/JS:   import { x } from 'module' / require()
//   Rust:    mod + use crate::path
//   Java:    import / import static; java.lang auto-imported
//   C++:     #include <header>
// ============================================================

import java.util.List;
import java.util.Map;
import java.util.HashMap;

import static java.lang.Math.PI;
import static java.lang.Math.sqrt;

public class Modules {

    static String greet(String name) {
        // java.lang.String is implicit — no import needed
        return "Hi";
    }

    public static void main(String[] args) {
        // ----- 1. Static import — PI -----
        System.out.printf("pi: %.2f%n", PI);                      // Output: pi: 3.14

        // ----- 2. java.lang implicit — String -----
        System.out.println("greet: " + greet("World"));            // Output: greet: Hi

        // ----- 3. Static import — sqrt -----
        System.out.println("sqrt: " + sqrt(25.0));                // Output: sqrt: 5.0

        // ----- 4. java.util.* — Map alias -----
        Map<String, Integer> scores = new HashMap<>();
        scores.put("alice", 25);
        System.out.println("alias: " + scores.get("alice"));      // Output: alias: 25

        // ----- 5. java.util List (from import java.util.List) -----
        List<String> tags = List.of("main", "module");
        System.out.println(tags.get(0) + " " + tags.get(1));      // Output: main module
    }
}